app/Models/Foundation/Summit/Registration/PromoCodes/Strategies/RegularPromoCodeTicketTypesStrategy.php

```diff
@@ -0,0 +1,154 @@
+<?php namespace App\Models\Foundation\Summit\Registration\PromoCodes\Strategies;
+/*
+ * Copyright 2023 OpenStack Foundation
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+use Illuminate\Support\Facades\Log;
+use models\main\Member;
+use models\summit\ISummitTicketType;
+use models\summit\Summit;
+use models\summit\SummitRegistrationPromoCode;
+use models\summit\SummitTicketType;
+use models\summit\SummitTicketTypeWithPromo;
+
+/**
+ * Class RegularPromoCodeTicketTypesStrategy
+ * @package App\Models\Foundation\Summit\Registration\PromoCodes\Strategies
+ */
+class RegularPromoCodeTicketTypesStrategy implements IPromoCodeAllowedTicketTypesStrategy
+{
+    /**
+     * @var Member
+     */
+    private $member;
+
+    /**
+     * @var SummitRegistrationPromoCode
+     */
+    protected $promo_code;
+
+    /**
+     * @var Summit
+     */
+    private $summit;
+
+    /**
+     * @param Summit $summit
+     * @param Member $member
+     * @param SummitRegistrationPromoCode|null $promo_code
+     */
+    public function __construct(Summit $summit, Member $member, ?SummitRegistrationPromoCode $promo_code)
+    {
+        $this->summit = $summit;
+        $this->promo_code = $promo_code;
+        $this->member = $member;
+    }
+
+    /**
+     * @param SummitTicketType $type
+     * @return ISummitTicketType
+     */
+    private function applyPromo2TicketType(SummitTicketType $type): ISummitTicketType {
+        if (!is_null($this->promo_code) && $this->promo_code->canBeAppliedTo($type)) {
+            return new SummitTicketTypeWithPromo($type, $this->promo_code);
+        }
+        return $type;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getTicketTypes(): array
+    {
+        $this->promo_code->checkSubject($this->member->getEmail(), $this->member->getCompany());
```

Comment:
promo code could be null
and u need to call validate here not checkSubject
and also need to catch any exception and set the promo code on null if there is any

Reply:
done